jsonrpc/server_pkg_test.go

```diff
@@ -0,0 +1,312 @@
+		reader := bytes.NewReader([]byte(test.req))
+		bufferedReader := bufio.NewReader(reader)
+		assert.Equal(t, test.batch, isBatch(bufferedReader), test.req)
+
+		left, err := io.ReadAll(bufferedReader)
+		assert.NoError(t, err)
+		assert.Equal(t, test.left, string(left))
+	}
+}
+
+func TestRequest(t *testing.T) {
+	failingTests := map[string]struct {
+		data []byte
+		want string
+	}{
+		"empty req":    {[]byte(""), "EOF"},
+		"empty object": {[]byte("{}"), "unsupported RPC request version"},
+		"wrong version": {[]byte(`
+				{
+					"jsonrpc" : "1.0"
+				}`), "unsupported RPC request version"},
+		"no method": {[]byte(`
+				{
+					"jsonrpc" : "2.0"
+				}`), "no method specified"},
+		"number param": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : 44
+				}`), "params should be an array or an object"},
+		"string param": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : "44"
+				}`), "params should be an array or an object"},
+		"array id": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : { "malatya" : "44"},
+					"id"     : [37]
+				}`), "id should be a string or an integer"},
+		"map id": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : { "malatya" : "44"},
+					"id"     : { "44" : "37"}
+				}`), "id should be a string or an integer"},
+		"float id": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : { "malatya" : "44"},
+					"id"     : 44.37
+				}`), "id should be a string or an integer"},
+	}
+
+	for desc, test := range failingTests {
+		_, err := newRequest(bytes.NewReader(test.data))
+		assert.Equal(t, err.Message, test.want, desc)
+	}
+
+	happyPathTests := map[string]struct {
+		data []byte
+	}{
+		"uint id": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : { "malatya" : "44"},
+					"id"     : 44
+				}`)},
+		"int id": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : { "malatya" : "44"},
+					"id"     : -44
+				}`)},
+		"string id": {[]byte(`
+				{
+					"jsonrpc" : "2.0",
+					"method" : "rpc_call",
+					"params" : { "malatya" : 44, "depdep" : { "plate" : "37" }},
+					"id"     : "-44"
+				}`)},
+	}
+
+	for desc, test := range happyPathTests {
+		_, err := newRequest(bytes.NewReader(test.data))
+		assert.Nil(t, err, desc)
+	}
+}
+
+func TestBuildArguments(t *testing.T) {
+	type arg struct {
+		Val1 int        `json:"val_1"`
+		Val2 string     `json:"val_2"`
+		Val3 *felt.Felt `json:"val_3"`
+	}
+
+	paramNames := []Parameter{{Name: "param1"}, {Name: "param2", Optional: true}}
+	felt, _ := new(felt.Felt).SetString("0x4437")
+
+	tests := []struct {
+		handler    any
+		req        string
+		shouldFail bool
+		errorMsg   string
+	}{
+		{
+			handler: func(param1 int, param2 *arg) (any, error) {
+				assert.Equal(t, 44, param1)
```

Comment:
Once you run these tests as subtests please make sure `t` refers to the correct test. Currently, `t` refers to the parent tests.